pkg/deploy/deployer.go

```diff
@@ -142,39 +206,116 @@ func (d *githubDeployer) Deploy(serviceNames []string, environmentName, commit,
 		}
 	}()
 
+	changesDetected := false
+
 	for service, environment := range serviceToEnvironment {
-		commitMsg := fmt.Sprintf("Deploy %s to %s with version %s triggered by %s (%s)", service.Name, environmentName, commit[:7], userFullname, userEmail)
-		files, err := d.renderTemplates(baseFolder, environment.TemplatePath, environment.GeneratedPath, service.Name, environmentName, commit)
-		if err != nil {
-			return nil, "", fmt.Errorf("failed to render templates for serivce %s, error: %w", service.Name, err)
+		commitMsg := fmt.Sprintf("%s %s to %s triggered by %s (%s)", action, service.Name, environment, userFullname, userEmail)
+		freezeFilePath := getFreezeFilePath(*environment)
+
+		var freezeFile string
+		if action == FreezeActionUnfreeze {
+			frozen, err := d.checkIfServiceFrozen(baseFolder, freezeFilePath)
+			if err != nil {
+				return nil, "", fmt.Errorf("failed to check if service %s is frozen, error: %w", service.Name, err)
+			}
+			if !frozen {
+				continue
+			}
+			freezeFile, err = d.removeFreezeFile(baseFolder, freezeFilePath)
+			if err != nil {
+				return nil, "", fmt.Errorf("failed to remove freeze file for service %s, error: %w", service.Name, err)
+			}
+		} else {
+			freezeFile, err = d.createFreezeFile(baseFolder, freezeFilePath)
+			if err != nil {
+				return nil, "", fmt.Errorf("failed to create freeze file for service %s, error: %w", service.Name, err)
+			}
 		}
 
-		tree, err := d.githubClient.CreateTree(ctx, ref, baseFolder, files)
+		tree, err := d.githubClient.CreateTree(ctx, ref, baseFolder, []string{freezeFile})
 		if err != nil {
-			return nil, "", fmt.Errorf("failed to create diff tree for serivce %s, error: %w", service.Name, err)
+			return nil, "", fmt.Errorf("failed to create diff tree for service %s, error: %w", service.Name, err)
 		}
 
 		if err = d.githubClient.PushCommit(ctx, ref, tree, userFullname, userEmail, commitMsg); err != nil {
-			return nil, "", fmt.Errorf("failed to create commit for serivce %s, error: %w", service.Name, err)
+			return nil, "", fmt.Errorf("failed to create commit for service %s, error: %w", service.Name, err)
 		}
+		changesDetected = true
 	}
 
-	prDescription := fmt.Sprintf("Service Names: %s\nEnvironment: %s\nCommit: [%s](%s)\nRequested by: %s (%s)",
-		servicesString, environmentName, commit[:7], commitUrl, userFullname, userEmail)
+	if !changesDetected {
+		logWithCtx.Info("No changes needed - services already in desired state")
+		return nil, "", nil
+	}
```

Comment:
will this case leave an open dangling branch?

Reply:
no, because this only will run when there was no commit (branch) pushed.